File: Leetcode 704 - binary-search/704.binary-search.cpp
/*
 * @lc app=leetcode id=704 lang=cpp
 *
 * [704] Binary Search
 */

// @lc code=start
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
  int search(vector<int> &nums, int target) {
    int low = 0;
    int high = nums.size() - 1;

    while (low <= high) {
      int mid = low + (high - low) / 2;
      if (nums[mid] == target) {
        return mid;
      } else if (nums[mid] < target) {
        low = mid + 1;
      } else {
        high = mid - 1;
      }
    }

    return -1;
  }
};

int main() {
  Solution sol;

  vector<int> nums = {1, 2, 3, 4, 5};
  int target = 3;
  cout << sol.search(nums, target) << endl; // 2

  vector<int> nums2 = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
  int target2 = 10;
  cout << sol.search(nums2, target2) << endl; // 9
}
// @lc code=end

File: Leetcode 704 - binary-search/704.binary-search.java
/*
 * @lc app=leetcode id=704 lang=java
 *
 * [704] Binary Search
 */

// @lc code=start
class Solution {

  public int search(int[] nums, int target) {
    int n = nums.length;
    int left = 0, right = n - 1;

    while (left <= right) {
      int mid = left + (right - left) / 2;

      if (nums[mid] == target) return mid;

      if (nums[mid] < target) {
        left = mid + 1;
      } else {
        right = mid - 1;
      }
    }

    return -1;
  }

  public static void main(String[] args) {
    Solution solution = new Solution();

    int[] nums = { -1, 0, 3, 5, 9, 12 };
    int target = 9;
    int result = solution.search(nums, target);
    System.out.println(result); // 4

    nums = new int[] { -1, 0, 3, 5, 9, 12 };
    target = 2;
    result = solution.search(nums, target);
    System.out.println(result); // -1
  }
}
// @lc code=end
